azure-ts-webserver-component/webserver.ts

```diff
@@ -0,0 +1,111 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as azure from "@pulumi/azure";
+import { specialArchiveSig } from "@pulumi/pulumi/runtime";
+
+// All resources will share a resource group.
+let resourceGroupName = new azure.core.ResourceGroup("server").name;
+
+// Create a network and subnet for all VMs.
+let network = new azure.network.VirtualNetwork("server-network", {
+    resourceGroupName,
+    addressSpaces: [ "10.0.0.0/16" ],
+    subnets: [{
+        name: "default",
+        addressPrefix: "10.0.1.0/24",
+    }],
+});
+let subnet = new azure.network.Subnet("server-subnet", {
+    resourceGroupName,
+    virtualNetworkName: network.name,
+    addressPrefix: "10.0.2.0/24",
+});
+
+/**
+ * WebServer is a reusable web server component that creates and exports a NIC, public IP, and VM.
+ */
+export class WebServer extends pulumi.ComponentResource {
+    public readonly publicIp: azure.network.PublicIp;
+    public readonly networkInterface: azure.network.NetworkInterface;
+    public readonly vm: azure.compute.VirtualMachine;
+
+    /**
+     * Allocate a new web server VM, NIC, and public IP address.
+     * @param name The name of the web server resource.
+     * @param args A bag of arguments to control the web server VM creation.
+     */
+    constructor(name: string, args: WebServerArgs) {
+        super("ws-ts-azure-comp:webserver:WebServer", name);
+
+        // Allocate a public IP and assign it to our NIC.
+        this.publicIp = new azure.network.PublicIp(`${name}-ip`, {
+            resourceGroupName,
+            allocationMethod: "Dynamic",
+        }, { parent: this });
+        this.networkInterface = new azure.network.NetworkInterface(`${name}-nic`, {
```

Comment:
we've been trying to move away from the pseudo hungarian `-xxx` suffixes for things.  I assume we're wildly inconsistent in examples.  But if we can prevent it from creeping in, we probably should.

Reply:
This is actually [recommended by Microsoft](https://docs.microsoft.com/en-us/azure/architecture/best-practices/naming-conventions):
> When naming resources in Azure, it is recommended to use common prefixes or suffixes to identify the type and context of the resource.
> Suggested Pattern
> [service short name]-[environment]-rg
> [vmname]-nic[num]

Any reason why you don't want to follow along?

Reply:
Apparently what's old is new again.  MS had hungarian, then moved away, and now moves back :)

If this is their recommendation, then this is totally fine.  

In general though, we've avoided this because it's generally redundant in pretty much any area we look in.  For example, if i'm on the AWS/Azure console then i'm enerally looking at resources that are already from some category.  i.e. i'm looking at my load-balancers.  So i don't need them to be suffixed with `-lb`.  Similarly, because a pulumi-urn already contains the type of the resource as part of it, it just ends up being a lot of redundancy.  i.e. if it already says `awsx:elb2:loadbalancer/...` i don't really need it to say `-lb` in the name to reaffirm that :)

But, like i said, if MS actively prefers this, then we should likely follow hte when-in-rome approach and accept that there.

Reply:
I don't have statistics, but it might be common to 
- look at a resource group with all sub-resources of different types
- search by name in the portal, and use resource type suffix while doing so

At least, I do so :)

Reply:
I guess this is a noop now that we go back to the same approach